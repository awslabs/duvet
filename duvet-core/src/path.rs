// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use core::fmt;
use serde::Deserialize;
use std::{ffi::OsStr, ops::Deref, path::PathBuf, sync::Arc};

#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Deserialize)]
#[serde(transparent)]
pub struct Path {
    path: Arc<OsStr>,
}

impl fmt::Debug for Path {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        self.as_ref().fmt(f)
    }
}

impl fmt::Display for Path {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        self.as_ref().display().fmt(f)
    }
}

impl Deref for Path {
    type Target = std::path::Path;

    fn deref(&self) -> &Self::Target {
        std::path::Path::new(&self.path)
    }
}

impl AsRef<std::path::Path> for Path {
    fn as_ref(&self) -> &std::path::Path {
        self
    }
}

impl PartialEq<str> for Path {
    fn eq(&self, other: &str) -> bool {
        self.as_ref().eq(std::path::Path::new(other))
    }
}

impl PartialEq<std::path::Path> for Path {
    fn eq(&self, other: &std::path::Path) -> bool {
        self.as_ref().eq(other)
    }
}

impl From<String> for Path {
    fn from(path: String) -> Self {
        Self {
            path: PathBuf::from(path).into_os_string().into(),
        }
    }
}

impl From<PathBuf> for Path {
    fn from(path: PathBuf) -> Self {
        Self {
            path: path.into_os_string().into(),
        }
    }
}

impl From<&Path> for Path {
    fn from(path: &Path) -> Self {
        Self {
            path: path.as_os_str().into(),
        }
    }
}

impl From<&str> for Path {
    fn from(path: &str) -> Self {
        Self {
            path: std::path::Path::new(path).as_os_str().into(),
        }
    }
}
