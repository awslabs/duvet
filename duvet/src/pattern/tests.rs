// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use super::*;
use std::path::{Path, PathBuf};

fn parse(pattern: &str, value: &str) -> Result<Vec<Annotation>, anyhow::Error> {
    let pattern = Pattern::from_arg(pattern).unwrap();
    let path = Path::new("file.rs");
    let mut annotations = Default::default();
    pattern.extract(value, path, &mut annotations)?;

    let annotations = annotations
        .into_iter()
        .map(|mut annotation| {
            // make the manifest dir consistent on all platforms
            annotation.manifest_dir = PathBuf::from("/");
            annotation
        })
        .collect();

    Ok(annotations)
}

macro_rules! snapshot {
    ($name:ident, $value:expr) => {
        snapshot!($name, "//=,//#", $value);
    };
    ($name:ident, $pattern:expr, $value:expr) => {
        #[test]
        fn $name() {
            insta::assert_debug_snapshot!(stringify!($name), parse($pattern, $value));
        }
    };
}

snapshot!(
    content_without_meta,
    r#"
    //# This is some content without meta
    "#
);

snapshot!(
    meta_without_content,
    r#"
    //= type=todo
    "#
);

snapshot!(
    type_citation,
    r#"
    //= https://example.com/spec.txt
    //# Here is my citation
    "#
);

snapshot!(
    type_test,
    r#"
    //= https://example.com/spec.txt
    //= type=test
    //# Here is my citation
    "#
);

snapshot!(
    type_todo,
    r#"
    //= https://example.com/spec.txt
    //= type=todo
    //= feature=cool-things
    //= tracking-issue=123
    //# Here is my citation
    "#
);

snapshot!(
    type_exception,
    r#"
    //= https://example.com/spec.txt
    //= type=exception
    //= reason=This isn't possible currently
    //# Here is my citation
    "#
);

snapshot!(
    missing_new_line,
    r#"
    //= https://example.com/spec.txt
    //# Here is my citation"#
);
