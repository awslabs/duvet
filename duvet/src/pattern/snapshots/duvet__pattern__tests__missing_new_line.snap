---
source: src/pattern/tests.rs
expression: "parse(\"//=,//#\",\n    r#\"\n    //= https://example.com/spec.txt\n    //# Here is my citation\"#)"
---
Ok(
    [
        Annotation {
            source: "file.rs",
            anno_line: 2,
            anno_column: 7,
            item_line: 3,
            item_column: 0,
            path: "",
            anno: Citation,
            target: "https://example.com/spec.txt",
            quote: "Here is my citation",
            comment: "",
            manifest_dir: "/",
            level: Auto,
            format: Auto,
            tracking_issue: "",
            feature: "",
            tags: {},
        },
    ],
)
